network/message_scope.go

```diff
@@ -0,0 +1,206 @@
+	return m.eventId[:]
+}
+
+func (m IncomingMessageScope) PayloadType() string {
+	return MessageType(m.msg.Payload)
+}
+
+// OutgoingMessageScope captures the context around an outgoing message that is about to be sent.
+type OutgoingMessageScope struct {
+	targetIds flow.IdentifierList               // the target node IDs.
+	channelId channels.Channel                  // the channel ID.
+	payload   interface{}                       // the payload to be sent.
+	encoder   func(interface{}) ([]byte, error) // the encoder to encode the payload.
+	msg       *message.Message                  // raw proto message sent on wire.
+	protocol  ProtocolType                      // the type of protocol used to send the message.
+}
+
+func NewOutgoingScope(
+	targetIds flow.IdentifierList,
+	channelId channels.Channel,
+	payload interface{},
+	encoder func(interface{}) ([]byte, error),
+	protocolType ProtocolType) (*OutgoingMessageScope, error) {
+	scope := &OutgoingMessageScope{
+		targetIds: targetIds,
+		channelId: channelId,
+		payload:   payload,
+		encoder:   encoder,
+		protocol:  protocolType,
+	}
+
+	if protocolType == ProtocolTypeUnicast {
+		// for unicast messages, we should have exactly one target.
+		if len(targetIds) != 1 {
+			return nil, fmt.Errorf("expected exactly one target id for unicast message, got: %d", len(targetIds))
+		}
+	}
+	if protocolType == ProtocolTypePubSub {
+		// for pubsub messages, we should have at least one target.
+		if len(targetIds) == 0 {
+			return nil, fmt.Errorf("expected at least one target id for pubsub message, got: %d", len(targetIds))
+		}
+	}
+
+	msg, err := scope.buildMessage()
+	if err != nil {
+		return nil, fmt.Errorf("could not build message: %w", err)
+	}
+	scope.msg = msg
+	return scope, nil
+}
+
+func (o OutgoingMessageScope) TargetIds() flow.IdentifierList {
+	return o.targetIds
+}
+
+func (o OutgoingMessageScope) Size() int {
+	return o.msg.Size()
+}
+
+func (o OutgoingMessageScope) PayloadType() string {
+	return MessageType(o.payload)
+}
+
+func (o OutgoingMessageScope) ChannelId() channels.Channel {
+	return o.channelId
+}
+
+// buildMessage builds the raw proto message to be sent on the wire.
+func (o OutgoingMessageScope) buildMessage() (*message.Message, error) {
+	payload, err := o.encoder(o.payload)
+	if err != nil {
+		return nil, fmt.Errorf("could not encode payload: %w", err)
+	}
+
+	emTargets := make([][]byte, 0)
+	for _, targetId := range o.targetIds {
+		tempID := targetId // avoid capturing loop variable
+		emTargets = append(emTargets, tempID[:])
+	}
+
+	return &message.Message{
+		TargetIDs: emTargets,
+		ChannelID: o.channelId.String(),
+		Payload:   payload,
+	}, nil
+}
+
+func (o OutgoingMessageScope) Proto() *message.Message {
+	return o.msg
+}
+
+// EventId computes the event ID for a given channel and payload (i.e., the hash of the payload and channel).
+func EventId(channel channels.Channel, payload []byte) (hash.Hash, error) {
```

Comment:
please document the return errors

Reply:
https://github.com/onflow/flow-go/pull/3675/commits/48c62f6999b464bec1742dca7fff9acba6344d9c